### ISSUE TITLE
Which envelope to inscribe on ETH ?

### ISSUE THREAD

**Issue by GaloisField2718** [ethscriptions-protocol/ESIPs#13]
Hello, 

I'm looking into ethinscription and I don't see any ˋOP_CODE` enveloppe to wrap inscription. 

Is it because I wrong read the doc or because it's not implemented nor think now ? 

Because basically, if it's only pushed in hex data on a transaction; it's like we are using ˋOP_RETURN` in Bitcoin. Especially we don't have any certification that data will be kept during other transfers. I mean, it wouldn't be wrapped into something easily transferable. And for all other purpose than brc20-like json protocol, it will be useless. 

I'm very curious to chat about it with you cause it seems to be important before to jump into this for me. 

Ordinals for Future 🦾

**Comment by RogerPodacter** [comment]
In Ethscriptions you aren't "transferring an ethscription" in the same way as I believe you are when you send a sat on Bitcoin.

Instead you "announce your desire to transfer it" (in a specific way specified by the protocol) and indexers reconstruct whether your desire resulted in a valid transfer, principally by verifying that you owned the ethscription when you tried to transfer it.

The "desire announcement paradigm" is more extensible because you can announce arbitrarily complicated things beyond just transfers, but it might have downsides also.

In any case, from a practical standpoint this shouldn't matter as the only relevant question is whether all state can be reconstructed by anyone using only decentralized data, and this is the case with both Ordinals and Ethscriptions.

**Comment by GaloisField2718** [comment]
> In Ethscriptions you aren't "transferring an ethscription" in the same way as I believe you are when you send a sat on Bitcoin.

Yes in fact on Bitcoin technically you are transferring script attached to the deployed transaction. As the minting process is tricky and interesting. 

> 
> Instead you "announce your desire to transfer it" (in a specific way specified by the protocol) and indexers reconstruct whether your desire resulted in a valid transfer, principally by verifying that you owned the ethscription when you tried to transfer it.
>
Ok so indeed this protocol is mainly for Json-based protocol on Ordinals. So we are mainly rebuilding not ordinals and inscriptions protocol but brc-like protocol. With a current global state updated by data pushed on the timechain. 

About this, did you seen XRC20 ? [Twitscritptions](https://app.twitscription.xyz/tokens). This showing for me that on any register we can index data and create tokens and marketplace. 

But, on Ordinals with inscriptions we can have Json-based protocols and other types of data. Which seems to not be the case in ethinscription. 


> In any case, from a practical standpoint this shouldn't matter as the only relevant question is whether all state can be reconstructed by anyone using only decentralized data, and this is the case with both Ordinals and Ethscriptions.

As long as it's only for json-based protocol yes it's the only important point. 
But new inscriptions are arriving with runes for example on ordinals. And this can change the way of making tokens protocol on Ordinals.

I'll try to stay tuned on ethinscription standard and evolution. Thanks a lot for your answer.

**Comment by RogerPodacter** [comment]
> Ok so indeed this protocol is mainly for Json-based protocol on Ordinals. So we are mainly rebuilding not ordinals and inscriptions protocol but brc-like protocol. With a current global state updated by data pushed on the timechain.

It's not just for JSON; you can ethscribe any content using a dataURI. Here's the first one ever: https://ethscriptions.com/ethscriptions/0

I believe Ethscriptions are functionally equivalent to Inscriptions, though the implementation is different.

Take a look at Facet (https://medium.com/@dumbnamenumbers/introducing-facet-5f9d43162117). It's a more general protocol built on top of Ethscriptions that abstracts away all implementation details and let's the user think in terms of traditional contracts.

**Comment by tunnckoCore** [comment]
> @RogerPodacter  In any case, from a practical standpoint this shouldn't matter as the only relevant question is whether all state can be reconstructed by anyone using only decentralized data, and this is the case with both Ordinals and Ethscriptions.

Practically, in fact, it's close to impossible (if not now, for sure in the future when we go massive) to reconstruct the whole thing alone, cuz it will mean you need to scan ALL transactions from day one, ALL of ALL of the dozens of millions of Ethereum transactions. That's my only and major concern. That's because the "transferring" is kinda "not part" of the protocol itself - meaning, if it was with `data:something/transfer,` it will be far easier to track and to reconstruct. For months i'm thinking to propose some ESIP for that.

> @GaloisField2718  Yes in fact on Bitcoin technically you are transferring script attached to the deployed transaction. As the minting process is tricky and interesting.

I don't think so. There you're not transferring anything too. They are UTXOs, and just the ownership is changed.

> @GaloisField2718 But, on Ordinals with inscriptions we can have Json-based protocols and other types of data. Which seems to not be the case in ethinscription.

That's not correct. You can push and host anything and everything. Both are absolutely the same. Consider transactions as satoshis, and you'll feel it. We can have "Rare sats" called "Rare transactions", because there are important and valueable ones, if the ethereum community come together and decide.

**Comment by GaloisField2718** [comment]
>That's because the "transferring" is kinda "not part" of the protocol itself - meaning, if it was with `data:something/transfer,` it will be far easier to track and to reconstruct. For months i'm thinking to propose some ESIP for that.

Definitely it could be a good idea to try!   

> I don't think so. There you're not transferring anything too. They are UTXOs, and just the ownership is changed.

In fact, UTXO are constantly destructed and rebuild. For me it's seems natural to think that is the PubScript attached to the original transaction which is shared. But indeed after many discussions, it's to be sure that we handle the same inscription over UTXOs, Casey purposed the sats' numbering. The Ordinal Theory allow us to track each sat and things inscribed on it, as if we destroy constantly each UTXOs. 
Maybe it's a wrong interpretation and it's a pleasure to discuss about it.

> That's not correct. You can push and host anything and everything. Both are absolutely the same. Consider transactions as satoshis, and you'll feel it. 

Yes indeed! Now it's very common. I made a mistake with this sentence, my bad.

> We can have "Rare sats" called "Rare transactions", because there are important and valueable ones, if the ethereum community come together and decide.

Yes, it's not the same scheme indeed. I don't know if the equivalent of sats are transactions but it couldn't be the `wei` for sure you're right. 

Still an open question for me to understand how we can implement such enveloppe `OP_IF` .... `OP_ENDIF` which can work to produce inscription in ordinals-like style.

**Comment by tunnckoCore** [comment]
> Yes, it's not the same scheme indeed. I don't know if the equivalent of sats are transactions but it couldn't be the wei for sure you're right.

Haha yea, that would be ideal but it's impossible cuz how the chain works - we can track weis but can track transactions.

> Still an open question for me to understand how we can implement such enveloppe OP_IF .... OP_ENDIF which can work to produce inscription in ordinals-like style.

About that.. don't know, but i actually played around a new protocol which i call Glyph (tokens oriented, calldata-based, highly efficient, faster, smaller, easier to index, and cheaper than even ethscritpions and facet), which kinda works with "OP codes". Like `0x1` mint, `0x2` transfer, and etc. It's more involved than that, but still far easier than JSON-based shits.
The reason it's highly efficient, and why i cannot do it ethscriptions-based, is cuz how i use the calldata (it's not hex decoded, but parsed).
